File: app/src/main/java/org/miaowo/miaowo/custom/ChatButton.java
package org.miaowo.miaowo.custom;

import android.content.Context;
import android.content.Intent;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.Point;
import android.graphics.PointF;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;

import org.miaowo.miaowo.R;
import org.miaowo.miaowo.activity.Chat;
import org.miaowo.miaowo.root.BaseActivity;
import org.miaowo.miaowo.util.SpUtil;

import java.util.Timer;
import java.util.TimerTask;

/**
 * 悬浮窗
 * Created by luqin on 17-1-21.
 */

public class ChatButton extends View {
    private static ChatButton chat;

    private Point mPosition = new Point();
    private Point mScreenSize = new Point();
    private PointF mStartTouch = new PointF();
    private PointF mEndTouch = new PointF();
    private PointF mChange = new PointF();
    private PointF mSize = new PointF(194, 194);
    private Paint mPaintFill;
    private WindowManager mManager;
    private SpUtil mSpUtil;
    private Drawable mLogo;
    private boolean mMoved = false;
    private Timer mMoveBack;

    public static void show(Context context) {
        if (chat == null) {
            chat = new ChatButton(context);
        }
        chat.show();
    }
    public static void hide() {
        chat.dismiss();
    }

    public ChatButton(Context context) {
        super(context);
        init();
    }
    public ChatButton(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }
    public ChatButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public void show() {
        WindowManager.LayoutParams params = buildLayoutParams();
        try {
            mManager.addView(this, params);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    private void dismiss() {
        try {
            mManager.removeViewImmediate(this);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    private WindowManager.LayoutParams buildLayoutParams() {
        WindowManager.LayoutParams params = new WindowManager.LayoutParams(
                WindowManager.LayoutParams.TYPE_APPLICATION_PANEL,
                WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
                        | WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL
                        | WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH,
                PixelFormat.RGBA_8888);
        params.gravity = Gravity.LEFT|Gravity.TOP;
        params.x = this.mPosition.x;
        params.y = this.mPosition.y;
        params.width = WindowManager.LayoutParams.WRAP_CONTENT;
        params.height = WindowManager.LayoutParams.WRAP_CONTENT;
        return params;
    }
    private void init() {
        mSpUtil = SpUtil.defaultSp();
        mManager = (WindowManager) getContext().getSystemService(Context.WINDOW_SERVICE);
        mManager.getDefaultDisplay().getSize(mScreenSize);
        mPosition.set(mSpUtil.getInt("btn_x", mScreenSize.x / 2), mSpUtil.getInt("btn_y", mScreenSize.y / 2));
        mPaintFill = new Paint();
        mPaintFill.setStyle(Paint.Style.FILL);
        int color = getResources().getColor(R.color.md_deep_purple_400);
        mPaintFill.setColor(Color.argb(100, Color.red(color), Color.green(color), Color.blue(color)));
        mLogo = getResources().getDrawable(R.drawable.chat);
    }

    private void move(int dx, int dy) {
        WindowManager.LayoutParams params = (WindowManager.LayoutParams) getLayoutParams();
        params.x += dx;
        params.y += dy;
        mPosition.set(params.x, params.y);
        mManager.updateViewLayout(this, params);
    }

    private void clickAction(int x, int y, boolean moved) {
        int xTo, yTo;
        if (!moved) {
            int finalY = y;
            if (x < 0 || x > mScreenSize.x - mSize.x) {
                if (x < 0) {
                    xTo = 0;
                    if (mMoveBack != null) {
                        mMoveBack.cancel();
                    }
                    mMoveBack = new Timer();
                    mMoveBack.schedule(new TimerTask() {
                        @Override
                        public void run() {
                            BaseActivity.get.runOnUiThreadIgnoreError(() -> clickAction(-1, finalY, true));
                            mMoveBack = null;
                        }
                    }, 600);
                } else {
                    xTo = (int) (mScreenSize.x - mSize.x);
                    if (mMoveBack != null) {
                        mMoveBack.cancel();
                    }
                    mMoveBack = new Timer();
                    mMoveBack.schedule(new TimerTask() {
                        @Override
                        public void run() {
                            BaseActivity.get.runOnUiThreadIgnoreError(() ->
                                    clickAction((int) (mScreenSize.x - mSize.x + 1), finalY, true));
                            mMoveBack = null;
                        }
                    }, 600);
                }
            } else {
                getContext().startActivity(new Intent(getContext(), Chat.class));
                return;
            }
        } else {
            xTo = x < 0 ? (int) (-mSize.x * 2 / 3) :
                    (int) (x > mScreenSize.x - mSize.x ? mScreenSize.x - mSize.x / 3 : x);
        }
        yTo = y < 0 ? 1 :
                (int) (y > mScreenSize.y - mSize.y ? mScreenSize.y - mSize.y : y);
        int dx = x > xTo ? -1 : x < xTo ? 1 : 0;
        int dy = y > yTo ? -1 : y < yTo ? 1 : 0;
        while (x != xTo || y != yTo) {
            move(dx, dy);
            x += dx;
            y += dy;
            dx = x > xTo ? -1 : x < xTo ? 1 : 0;
            dy = y > yTo ? -1 : y < yTo ? 1 : 0;
        }
        mPosition.x = xTo;
        mPosition.y = yTo;
        mSpUtil.putInt("btn_x", mPosition.x);
        mSpUtil.putInt("btn_y", mPosition.y);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        mSize.set(getMeasuredWidth(), getMeasuredHeight());
        canvas.drawCircle(mSize.x / 2, mSize.y / 2, Math.min(mSize.x, mSize.y) / 2, mPaintFill);
        mLogo.setBounds((int) (mSize.x / 4), (int) (mSize.y / 4)
                , (int) (mSize.x * 3 / 4), (int) (mSize.y * 3 / 4));
        mLogo.draw(canvas);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        setMeasuredDimension(MeasureSpec.getMode(widthMeasureSpec) == MeasureSpec.EXACTLY
                        ? MeasureSpec.getSize(widthMeasureSpec) : (int) mSize.x,
                MeasureSpec.getMode(heightMeasureSpec) == MeasureSpec.EXACTLY
                        ? MeasureSpec.getSize(heightMeasureSpec) : (int) mSize.y);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                mStartTouch.set(event.getX(), event.getY());
                mMoved = false;
                return true;
            case MotionEvent.ACTION_MOVE:
                mEndTouch.set(event.getX(), event.getY());
                mChange.set(mEndTouch.x - mStartTouch.x, mEndTouch.y - mStartTouch.y);
                if (Math.sqrt(Math.pow(mChange.x, 2) + Math.pow(mChange.y, 2)) >= 3) {
                    move((int) mChange.x, (int) mChange.y);
                    mMoved = true;
                }
                return true;
            case MotionEvent.ACTION_UP:
                clickAction(mPosition.x, mPosition.y, mMoved);
                return true;
        }
        return false;
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        BaseActivity.get.onKeyDown(keyCode, event);
        return true;
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        BaseActivity.get.onKeyUp(keyCode, event);
        return true;
    }
}

File: app/src/main/java/org/miaowo/miaowo/ui/ChatButton.kt
package org.miaowo.miaowo.ui

import android.content.Context
import android.content.Intent
import android.graphics.*
import android.graphics.drawable.Drawable
import android.support.v4.content.res.ResourcesCompat
import android.util.AttributeSet
import android.view.*
import org.miaowo.miaowo.R
import org.miaowo.miaowo.activity.Chat
import org.miaowo.miaowo.base.App
import org.miaowo.miaowo.base.extra.activity
import org.miaowo.miaowo.base.extra.uiThread
import org.miaowo.miaowo.base.extra.spGet
import org.miaowo.miaowo.base.extra.spPut
import java.util.*
import kotlin.properties.Delegates

/**
 * 悬浮窗
 * Created by luqin on 17-1-21.
 */

class ChatButton : View {

    private val mPosition = Point()
    private val mScreenSize = Point()
    private val mStartTouch = PointF()
    private val mEndTouch = PointF()
    private val mChange = PointF()
    private var mPaintFill = Paint()
    private var mManager: WindowManager? = null
    private var mLogo: Drawable? = null
    private var mSize by Delegates.notNull<Float>()
    private var mMoved = false
    private var mMoveBack: Timer? = null

    private fun init() {
        mManager = activity?.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        mManager?.defaultDisplay?.getSize(mScreenSize)
        mSize = (Math.max(mScreenSize.x, mScreenSize.y) / 32 * 3).toFloat()
        mPosition.set(spGet("btn_x", mScreenSize.x / 2), spGet("btn_y", mScreenSize.y / 2))
        mPaintFill.style = Paint.Style.FILL
        val color = ResourcesCompat.getColor(resources, R.color.md_deep_purple_400, null)
        mPaintFill.color = Color.argb(100, Color.red(color), Color.green(color), Color.blue(color))
        mLogo = ResourcesCompat.getDrawable(resources, R.drawable.chat, null)
    }

    constructor(context: Context) : super(context) { init() }
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) { init() }
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) { init() }

    fun show() {
        val params = buildLayoutParams()
        try {
            mManager!!.addView(this, params)
        }
        catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun dismiss() {
        try { mManager?.removeViewImmediate(this) }
        catch (e: Exception) { e.printStackTrace() }
    }

    private fun buildLayoutParams(): WindowManager.LayoutParams {
        val params = WindowManager.LayoutParams(
                WindowManager.LayoutParams.TYPE_APPLICATION_PANEL,
                WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS or WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL or WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH,
                PixelFormat.RGBA_8888)
        params.gravity = Gravity.START or Gravity.TOP
        params.x = this.mPosition.x
        params.y = this.mPosition.y
        params.width = WindowManager.LayoutParams.WRAP_CONTENT
        params.height = WindowManager.LayoutParams.WRAP_CONTENT
        return params
    }

    private fun move(dx: Int, dy: Int) {
        val params = layoutParams as WindowManager.LayoutParams
        params.x += dx
        params.y += dy
        mPosition.set(params.x, params.y)
        mManager!!.updateViewLayout(this, params)
    }

    private fun clickAction(x: Int, y: Int, moved: Boolean) {
        var aX = x
        var aY = y
        val xTo: Int
        val yTo: Int
        if (!moved) {
            val finalY = aY
            if (aX < 0 || aX > mScreenSize.x - mSize) {
                if (aX < 0) {
                    xTo = 0
                    mMoveBack?.cancel()
                    mMoveBack = Timer()
                    mMoveBack?.schedule(object : TimerTask() {
                        override fun run() {
                            activity?.uiThread { clickAction(-1, finalY, true) }
                            mMoveBack = null
                        }
                    }, 600)
                } else {
                    xTo = (mScreenSize.x - mSize).toInt()
                    mMoveBack?.cancel()
                    mMoveBack = Timer()
                    mMoveBack?.schedule(object : TimerTask() {
                        override fun run() {
                            activity?.uiThread { clickAction((mScreenSize.x - mSize + 1).toInt(), finalY, true) }
                            mMoveBack = null
                        }
                    }, 600)
                }
            } else {
                context.startActivity(Intent(context, Chat::class.java))
                return
            }
        } else {
            xTo = if (aX < 0) (-mSize * 2 / 3).toInt()
            else if (aX > mScreenSize.x - mSize) (mScreenSize.x - mSize / 3).toInt() else aX
        }
        yTo = if (aY < 0) 1
        else if (aY > mScreenSize.y - mSize) (mScreenSize.y - mSize).toInt() else aY
        var dx = if (aX > xTo) -1 else if (aX < xTo) 1 else 0
        var dy = if (aY > yTo) -1 else if (aY < yTo) 1 else 0
        while (aX != xTo || aY != yTo) {
            move(dx, dy)
            aX += dx
            aY += dy
            dx = if (aX > xTo) -1 else if (aX < xTo) 1 else 0
            dy = if (aY > yTo) -1 else if (aY < yTo) 1 else 0
        }
        mPosition.x = xTo
        mPosition.y = yTo
        spPut("btn_x", mPosition.x)
        spPut("btn_y", mPosition.y)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        mSize = measuredWidth.toFloat()
        canvas.drawCircle(mSize / 2, mSize / 2, Math.min(mSize, mSize) / 2, mPaintFill)
        mLogo?.setBounds((mSize / 4).toInt(), (mSize / 4).toInt(), (mSize * 3 / 4).toInt(), (mSize * 3 / 4).toInt())
        mLogo?.draw(canvas)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        setMeasuredDimension(if (View.MeasureSpec.getMode(widthMeasureSpec) == View.MeasureSpec.EXACTLY)
            View.MeasureSpec.getSize(widthMeasureSpec)
        else
            mSize.toInt(),
                if (View.MeasureSpec.getMode(heightMeasureSpec) == View.MeasureSpec.EXACTLY)
                    View.MeasureSpec.getSize(heightMeasureSpec)
                else
                    mSize.toInt())
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                mStartTouch.set(event.x, event.y)
                mMoved = false
                return true
            }
            MotionEvent.ACTION_MOVE -> {
                mEndTouch.set(event.x, event.y)
                mChange.set(mEndTouch.x - mStartTouch.x, mEndTouch.y - mStartTouch.y)
                if (Math.sqrt(Math.pow(mChange.x.toDouble(), 2.0) + Math.pow(mChange.y.toDouble(), 2.0)) >= 3) {
                    move(mChange.x.toInt(), mChange.y.toInt())
                    mMoved = true
                }
                return true
            }
            MotionEvent.ACTION_UP -> {
                clickAction(mPosition.x, mPosition.y, mMoved)
                return true
            }
        }
        return false
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        activity?.onKeyDown(keyCode, event)
        return true
    }

    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
        activity?.onKeyUp(keyCode, event)
        return true
    }

    companion object {
        private var chat: ChatButton? = null

        fun show() {
            if (chat == null) {
                chat = ChatButton(App.i)
            }
            chat?.show()
        }

        fun hide() = chat?.dismiss()
    }
}
